aws-api/src/main/java/com/amplifyframework/api/aws/sigv4/AppSyncSigV4SignerInterceptor.java

```diff
@@ -58,62 +60,92 @@
     private final OidcAuthProvider oidcAuthProvider;
     private final String awsRegion;
     private final AuthorizationType authType;
+    private final EndpointType endpointType;
+
+    private AppSyncSigV4SignerInterceptor(AWSCredentialsProvider credentialsProvider,
+                                          ApiKeyAuthProvider apiKeyProvider,
+                                          CognitoUserPoolsAuthProvider cognitoUserPoolsAuthProvider,
+                                          OidcAuthProvider oidcAuthProvider,
+                                          String awsRegion,
+                                          AuthorizationType authType,
+                                          EndpointType endpointType) {
+        this.credentialsProvider = credentialsProvider;
+        this.apiKeyProvider = apiKeyProvider;
+        this.cognitoUserPoolsAuthProvider = cognitoUserPoolsAuthProvider;
+        this.oidcAuthProvider = oidcAuthProvider;
+        this.awsRegion = awsRegion;
+        this.authType = authType;
+        this.endpointType = endpointType;
+    }
 
     /**
      * Constructs an instance of AppSyncSigV4SignerInterceptor that
      * uses API key for authorization.
      * @param apiKeyProvider An instance of {@link ApiKeyAuthProvider}
+     * @param endpointType Endpoint type for the api
      */
-    public AppSyncSigV4SignerInterceptor(@NonNull ApiKeyAuthProvider apiKeyProvider) {
-        this.apiKeyProvider = Objects.requireNonNull(apiKeyProvider);
-        this.credentialsProvider = null;
-        this.cognitoUserPoolsAuthProvider = null;
-        this.oidcAuthProvider = null;
-        this.awsRegion = null;
-        this.authType = AuthorizationType.API_KEY;
+    public AppSyncSigV4SignerInterceptor(@NonNull ApiKeyAuthProvider apiKeyProvider,
```

Comment:
A builder would be better for this class.

Reply:
I agree ... the `AppSyncSigV4SignerInterceptor` is tech debt that we carried over from the App Sync SDK. At the time we brought it in, it was argued "we won't modify it, we'll just leave it like it was in the App Sync code base."

If we are going to modify it, we should pretty much re-write it. They way it constructs, with varying groups of `null` params, is fairly insane.

I think what we should do is have 1 interceptor implementation per auth type. Then, maybe there is some factory that sits above them, and can select an appropriate interceptor, based on the configured auth type. But all of the details for each auth type would be delegated to their respective interceptor implementations.

Reply:
@jamesonwilliams these are good points, let's do this in a single, focused refactor PR.